### ISSUE TITLE
Support skipping parts of message template when fields are undefined

### ISSUE THREAD

**Issue by hosiet** [systemd/systemd#3351]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> 221+ (ever since catalog exists)
> tested on 229

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> Debian testing
> Arch Linux
### In case of bug report: Expected behaviour you didn't see

```
> # LC_ALL=C journalctl --catalog
> ...(misc messages skipped)
> May 26 18:23:10 debian-s400ca systemd[1301]: Startup finished in 43ms.
-- Subject: System start-up is now complete
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- All system services necessary queued for starting at boot have been
-- successfully started. Note that this does not mean that the machine is
-- now idle as services might still be busy with completing start-up.
-- 
-- Kernel start-up required 12345 microseconds.
-- 
-- Initial RAM disk start-up required 32127 microseconds.
-- 
-- Userspace start-up required 43712 microseconds.
```
### In case of bug report: Unexpected behaviour you saw

```
> # LC_ALL=C journalctl --catalog
> ...(misc messages skipped)
> May 26 18:23:10 debian-s400ca systemd[1301]: Startup finished in 43ms.
> -- Subject: System start-up is now complete
> -- Defined-By: systemd
> -- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
> -- 
> -- All system services necessary queued for starting at boot have been
> -- successfully started. Note that this does not mean that the machine is
>-- now idle as services might still be busy with completing start-up.
> -- 
> -- Kernel start-up required KERNEL_USEC microseconds.
> -- 
> -- Initial RAM disk start-up required INITRD_USEC microseconds.
> -- 
> -- Userspace start-up required 43712 microseconds.
```
### In case of bug report: Steps to reproduce the problem

> # LC_ALL=C journalctl --catalog

---

P.S. There are also other bad catalogs, e.g.,

```
May 26 18:22:45 debian-s400ca systemd[1127]: Reached target Basic System.
-- Subject: Unit UNIT has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished starting up.
-- 
-- The start-up result is done.
May 26 18:22:45 debian-s400ca systemd[1127]: Reached target Default.
-- Subject: Unit UNIT has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished starting up.
-- 
-- The start-up result is done.
```

**Comment by evverx** [comment]
https://www.freedesktop.org/wiki/Software/systemd/catalog/

> When a catalog entry is printed on screen for a specific log entry simple variable replacements are applied. Journal field names enclosed in @ will be replaced by their values, if such a field is available in an entry. If such a field is not defined in an entry the enclosing @ will be dropped but the variable name is kept.

Try `sudo journalctl -b MESSAGE_ID=b07a249cd024414a82dd00cd181378ff -o verbose`.
Some messages don't contain `KERNEL_USEC` or/and `INITRD_USEC` (and that's ok for `systemd --user`)

**Comment by evverx** [comment]
> P.S. There are also other bad catalogs

This seems like a bug. Indeed, there is no `UNIT`-field. But there is a `USER_UNIT`-field.

**Comment by keszybz** [comment]
Writing catalog entries for messages which have a non-constant set of fields is currently impossible. The obvious solution is to create a different MESSAGE_ID for each case, and then provide separate catalog entries. But this doesn't really scale; for example we'd need three different MESSAGE_IDs for SD_MESSAGE_STARTUP_FINISHED; the catalog gets bigger and contains the same repeated text with small variations, and most importantly matching messages becomes annoying. In the end, I don't think this is a good solution. It would seem much nicer to allow some flexibility in how catalog messages are parsed:

I have the following ideas:
1. allow "field fallback" by specifying a list of fields to match `@FIELD:FALLBACK_FIELD_1[:FALLBACK_FIELD_2]@`. This would work for the following case:
```
    Unit @UNIT:USER_UNIT@ has finished starting up.
```

2. allow "field conditionals", i.e. provide part of the text only if some field is present:
```
All system services necessary queued for starting at boot have been
started. Note that this does not mean that the machine is now idle as services
might still be busy with completing start-up.

{KERNEL_USEC?Kernel start-up required @KERNEL_USEC@ microseconds.}

{INITRD_USEC?Initial RAM disk start-up required @INITRD_USEC@ microseconds.}

{USERSPACE_USEC?Userspace start-up required @USERSPACE_USEC@ microseconds.}
```
(double empty lines would be suppressed.)

Comments?

**Comment by poettering** [comment]
I am really afraid of inventing a full-blown macro language for this I must say, but I don't see how we could solve this otherwise I must admit...

The fallback syntax looks good though and simple enough. Matches `$PATH` in some ways, i.e. syntax and "first found" logic, hence makes a lot of sense to me.

The syntax for the conditionals bit I am not sure I like too much I must say... Maybe instead we should declare that paragraphs are suppressed automatically if they reference a variable that doesn't exist? Would be less flexible, but I like the appeal of not having to extend the syntax at all for it? Maybe we could even permit `@@@` as a special token where additional virtual paragraph breaks would be placed when doing the replacements, without actually placing a real one? Example:

```
This sentence is visible in all cases. @@@And this sentence is only visible if the variable @FOOBAR@ is set, but is otherwise surprised, starting from the previous virtual break at the three at-signs, and all the way to the ones following here: @@@And this is shown always again @@@ and this only if @WALDO@ is set, @@@ maybe extended by this when @QUUX@ and @YIPPIEH@ are set.
```

The above would be rendered like so, if WALDO=miep and QUUX=wuff are set:

```
This sentence is visible in all cases. And this is shown always again and this only if miep is set, 
```

If you follow what I mean?

(Not sure I like the precise syntax and semantics of my proposal above, but I do like that it doesn't introduce an antirely new control character and is a very natural extension I think. Dunno. Ideas, maybe based on this one?

**Comment by keszybz** [comment]
Yeah, I though about simply skipping paragraphs too. That would work too. INevertheless, I also don't think we should silently start skipping paragraphs which are missing some term. In particular, when structured messages change over time, it's rather common to add fields, and then the catalog entry stop matching old entries. It's better to display the message with the format string substituted, than totally skip the paragraph. So I think this skipping should be performed only when explicitly requested.

```
I'm a sentence that is always shown @UNDEFINED_FIELD@.

@@I'm a sentence that is always skipped @UNDEFINED_FIELD@.

I'm always shown. @@I'm always skipped @UNDEFINED_FIELD@.

I'm always shown. @@I'm shown if @FOOBAR@ is defined.
```

**Comment by poettering** [comment]
So you'd propose the rule: "if `@@` has already appeared in a paragraph and we encounter a substitution we cannot execute, drop the everything starting with that previous `@@` until the next `@@` or the end of the paragraph, whatever comes next."?

How would we define "paragraph"? the next newline? or the next double newline?

(I probably wouldn't use `@@` for this, I figure we should leave that for placing a single `@` character in the text, without replacing. Which is why I used `@@@` in my proposal, to mark such cases.)

**Comment by keszybz** [comment]
> So you'd propose the rule: "if @@ has already appeared in a paragraph and we encounter a substitution we cannot execute, drop the everything starting with that previous @@ until the next @@ or the end of the paragraph, whatever comes next."?

Exactly.

> How would we define "paragraph"? the next newline? or the next double newline?

Double new-line. Catalog texts are (at least ours) wrapped, so a single newline is not a useful distinction.

**Comment by poettering** [comment]
Hmm, but that would create a slight asymmtry: at a beginning of a paragraph it would be possible to use the "replace empty var by empty string" logic, until the first ``@@@` is reached — but at the end of a paragraph you couldn't return to this mode. Maybe, in order to make this symmetric, change the rule to this:

> When we encounter an unresolvable variable, look for the previous `@@@` and the next `@@@` within the current paragraph. If both are found, suppress the whole substring between them. If one or both of them are missing, replace the variable by the empty string. 

Hence:

```
This will always stay, even if @FOOBAR@ cannot be resolved.@@@ But this is entirely suppressed
if @QUUX@ cannot be resolved.@@@ And this will always stay again, even if @WALDO@ cannot
be replaced.
```

Of course, one couldn't switch back to the replace-by-empty mode in the middle of a paragraph, but I think that's fine.

(Double new-line sounds fine to me.)

**Comment by poettering** [comment]
Btw there's also #1972...  i figure we can close one of these two?
